client/components/data/query-themes/index.jsx

```diff
@@ -0,0 +1,54 @@
+/**
+ * External dependencies
+ */
+import { Component, PropTypes } from 'react';
+import shallowEqual from 'react-pure-render/shallowEqual';
+import { connect } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { requestThemes } from 'state/themes/actions';
+import { isRequestingThemesForQuery } from 'state/themes/selectors';
+
+class QueryThemes extends Component {
+	static propTypes = {
+		siteId: PropTypes.oneOfType( [
+			PropTypes.number,
+			PropTypes.oneOf( [ 'wpcom' ] )
+		] ).isRequired,
+		query: PropTypes.object,
+		// Connected props
+		isRequesting: PropTypes.bool.isRequired,
+		requestThemes: PropTypes.func.isRequired,
+	}
+
+	componentDidMount() {
+		this.request( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
```

Comment:
I think we can simplify things a bit by using `shouldComponentUpdate` instead of `componentWillReceiveProps`. If we do that then we don't need to pass different props into the `request` method.

we can just do 

```jsx
shouldComponentUpdate( nextProps ) {
    return this.props.siteId !== nextProps.siteId || ! shallowEqual( this.props.query, nextProps.query );
}

componentDidMount() {
    this.request();
}

componentDidUpdate() {
    this.request();
}
```